lib/collection/src/update_handler.rs

```diff
@@ -401,12 +461,46 @@ impl UpdateHandler {
                     {
                         continue;
                     }
+
+                    // Continue if we have enough CPU budget available to start an optimization
+                    // Otherwise skip now and start a task to trigger the optimizer again once CPU
+                    // budget becomes available
+                    let desired_cpus = num_rayon_threads(max_indexing_threads);
+                    if !optimizer_cpu_budget.has_budget(desired_cpus) {
+                        let trigger_active = cpu_available_trigger
+                            .as_ref()
+                            .map_or(false, |t| !t.is_finished());
+                        if !trigger_active {
+                            cpu_available_trigger.replace(trigger_optimizers_on_cpu_budget(
```

Comment:
I would say, this function is not exactly required: if we don't have CPU budget to run single optimization, that would mean that at least one another optimization is already running. When optimization finishes - it generates exactly same signal as `trigger_optimizers_on_cpu_budget`. 

On the other hand, having this function doesn't hurt either, cause then we can avoid relying on external logic.
And even if it will trigger one extra signal, it won't create any more computations.

Reply:
> I would say, this function is not exactly required: if we don't have CPU budget to run single optimization, that would mean that at least one another optimization is already running. When optimization finishes - it generates exactly same signal as trigger_optimizers_on_cpu_budget.

I thought this too, at first, but this is not the case.

It assumes that we only have one shard in which we do optimizations, with one instance of CPU budget.

This becomes a problem if you have multiple shards, because these share CPU budget (as you only have one SoC). Imagine that you have two shards. Shard A claimed all CPU budget, shard B will not start optimization because there is no CPU budget left. If optimizations on shard A finish, we don't trigger shard B. B will then only be triggered by external factors, such as upsertions.

That's the only reason for having this logic here. I'd love to remove it though, but I don't see a better way of doing things.